Lesson_20/oop-principles.js

```diff
@@ -0,0 +1,202 @@
+document.addEventListener('DOMContentLoaded', () => {
+   class Human {
+      constructor(data) {
+         this.name = data.name;
+         this.surname = data.surname;
+         this.age = data.age;
+      }
+
+      getFullName() {
+         return `${this.name} ${this.surname}`;
+      }
+
+      setFullName(fullName) {
+         [this.name, this.surname] = fullName.split(' ');
+      }
+   }
+
+   const avgMark = (arr) => {
+      const sumOfItem = (arr) => arr.reduce((acc, el) => acc + el, 0);
+      return sumOfItem(arr) / arr.length;
+   }
+
+   class Teacher extends Human {
+      constructor(data) {
+         super(data);
+         this.group = data.group;
+      }
+
+      getListOfNamesByAverageMark() {
+         const sortbyAvMark = () => (a, b) => a.getAverageMark() < b.getAverageMark() ? 1 : -1;
+         const copyGroup = [...this.group];
+         return copyGroup.sort(sortbyAvMark()).map(student => student.name);
+      }
+      
+      getStudentByName(name) {
+         const filterByName = this.group.filter(student => student.name === name);
+         return filterByName[0];
+      }
+
+      #studentIndexByName(name) {
+         return this.group.indexOf(this.getStudentByName(name));
+      }
+
+      removeStudentByName(name) {
+         this.group.splice(this.#studentIndexByName(name), 1);
+      }
+
+      updateStudentByName(student, name) {
+         this.group.splice(this.#studentIndexByName(name), 1, student);
+      }
+
+      findFakeStudentByProto() {
+         let fakeStudent;
+         this.group.forEach(student => {
+            if(Object.getPrototypeOf(student) !== Human) {
+               fakeStudent = student;
+            }
+         });
+         return `${fakeStudent.name} has original marks: ${fakeStudent.marks}`;
+      }
+
+      findFakeStudentByAverageMark() {
+         let fakeStudent;
+         this.group.forEach(student => {
+            if(student.getAverageMark() !== avgMark(student.marks)) {
+               fakeStudent = student;
+            }
+         })
+         return `${fakeStudent.name} has original marks: ${fakeStudent.marks}`;
+      }
+   }
+
+   class Student extends Human {
+      constructor(data) {
+         super(data);
+         this.marks = data.marks;
+      }
+
+      getAverageMark() {
+         return avgMark(this.marks);
+      }
+
+      getMinMark() {
+         return Math.min(...this.marks);
+      }
+
+      getMaxMark() {
+         return Math.max(...this.marks);
+      }
+   }
+
+   class FakeStudent extends Student {
+      #cheatedMarks = this.#cheat();
+
+      #cheat() {
+         const maxMark = 12;
+         const preferMark = 11;
+         const multiplier = 2;
+
+         const setMark = (mark) => {
+            const multMark = mark * multiplier;
+            return multMark >= maxMark ? preferMark : multMark;
```

Comment:
Why do you need `preferMark`? Accordingly to hometask requirements it should be `maxMark`